browser/ephemeral_storage/ephemeral_storage_tab_helper.cc

```diff
@@ -0,0 +1,183 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ephemeral_storage/ephemeral_storage_tab_helper.h"
+
+#include <set>
+#include <string>
+#include <utility>
+
+#include "base/feature_list.h"
+#include "base/no_destructor.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/ui/browser.h"
+#include "chrome/browser/ui/browser_list.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/session_storage_namespace.h"
+#include "content/public/browser/storage_partition.h"
+#include "content/public/browser/web_contents.h"
+#include "net/base/registry_controlled_domains/registry_controlled_domain.h"
+#include "third_party/blink/public/common/features.h"
+
+#if defined(OS_ANDROID)
+#include "chrome/browser/android/tab_android.h"
+#include "chrome/browser/ui/android/tab_model/tab_model_list.h"
+#endif
+
+using content::BrowserContext;
+using content::NavigationHandle;
+using content::WebContents;
+
+namespace {
+
+content::SessionStorageNamespaceMap& session_storage_namespace_map() {
+  static base::NoDestructor<content::SessionStorageNamespaceMap>
+      session_storage_namespace_map;
+  return *session_storage_namespace_map.get();
+}
+
+content::SessionStorageNamespaceMap& local_storage_namespace_map() {
+  static base::NoDestructor<content::SessionStorageNamespaceMap>
+      local_storage_namespace_map;
+  return *local_storage_namespace_map.get();
+}
+
+std::string URLToStorageDomain(const GURL& url) {
+  std::string domain = net::registry_controlled_domains::GetDomainAndRegistry(
+      url, net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES);
+
+  // GetDomainAndRegistry might return an empty string if this host is an IP
+  // address or a file URL.
+  if (domain.empty())
+    domain = url::Origin::Create(url.GetOrigin()).Serialize();
+
+  return domain;
+}
+
+}  // namespace
+namespace ephemeral_storage {
+
+EphemeralStorageTabHelper::~EphemeralStorageTabHelper() {}
+
+EphemeralStorageTabHelper::EphemeralStorageTabHelper(WebContents* web_contents)
+    : WebContentsObserver(web_contents) {}
+
+void EphemeralStorageTabHelper::ReadyToCommitNavigation(
+    NavigationHandle* navigation_handle) {
+  if (!base::FeatureList::IsEnabled(blink::features::kBraveEphemeralStorage))
+    return;
+  if (!navigation_handle->IsInMainFrame())
+    return;
+  if (navigation_handle->IsSameDocument())
+    return;
+
+  std::string domain = URLToStorageDomain(navigation_handle->GetURL());
+  std::string previous_domain =
+      URLToStorageDomain(web_contents()->GetLastCommittedURL());
+  if (domain == previous_domain)
+    return;
+
+  ClearEphemeralStorageIfNecessary();
+
+  auto* browser_context = web_contents()->GetBrowserContext();
+  auto site_instance = content::SiteInstance::CreateForURL(
```

Comment:
this is not necessary if we wanted just `BrowserContext::GetDefaultStoragePartition`

Reply:
Hrm. This properly calls `GetStoragePartitionConfigForSite`. Perhaps that only matters for plugins though.